streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -655,10 +656,11 @@ public synchronized void onChange(final Thread thread,
                     // global stream thread has different invariants
                     final GlobalStreamThread.State newState = (GlobalStreamThread.State) abstractNewState;
                     globalThreadState = newState;
+                    final GlobalStreamThread.State oldState = (GlobalStreamThread.State) abstractOldState;
 
                     if (newState == GlobalStreamThread.State.RUNNING) {
                         maybeSetRunning();
-                    } else if (newState == GlobalStreamThread.State.DEAD) {
+                    } else if (newState == GlobalStreamThread.State.DEAD && oldState != GlobalStreamThread.State.PENDING_SHUTDOWN) {
```

Comment:
Why does this lead to an error state?
AFAICS, an exceptional shutdown of the global stream thread would also first put the global stream thread into `PENDING_SHUTDOWN` and then `DEAD` (see `catch` clauses in `run()` in `GlobalStreamThread`).
In any case, tests are needed to verify this code.

Reply:
Actually, a transition to `DEAD` from a different state than `PENDING_SHUTDOWN` is illegal according to the state machine in global stream thread.

Reply:
Not sure what you are trying to say? Can you clarify?

I agree that we always need to be in `PENDING_SHUTDOWN` to be able to transit to `DEAD`. That's why it would be an error the we should log.

Not sure if you agree to the propose code change or not?

Reply:
My point is that if the transition from a different state than `PENDING_SHUTDOWN` is illegal then we will never reach the `log.error()` on line 664 since [this check](https://github.com/apache/kafka/blob/c1071327c50613ae8647f1fa98f00f78a0cb5d76/streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java#L163) will throw an exception before.